### PR TITLE
Add D-FINE Decoder

### PR DESCRIPTION
**Pull request by yutong-xiang-97**
## What has changed and why?

Add D-FINE decoder based on the [original implementation](https://github.com/Peterande/D-FINE/blob/master/src/zoo/dfine/dfine_decoder.py). Meanwhile, we retain backward compatibility with using the old RTDETRv2 decoder with a configurable `model_args` entry.

Major changes include (based on D-FINE paper):
- replacing original decoder projection layer with a Target Gating Layer
- applying Fine-grained Distribution Refinement (FDR)

The `dfine_decoder.py` and `dfine_utils.py` are mostly carbon copies of the original except for:
- adding ``score_head_reuse_or_reinit_hook`` and (when denoising is enabled) ``denoising_class_embed_reuse_or_reinit_hook`` as `load_state_dict` pre-hooks, so pretrained D-FINE checkpoints can be adapted to a new ``num_classes`` at load time. The same hooks are refactored from the `rtdetrv2_decoder.py` to a separate `hooks.py`.
- registering ``anchors`` / ``valid_mask`` for evaluation as non-persistent buffers instead so they are not written into checkpoints.
- making ``value_op`` return the flat ``[bs, L, n_head, c]`` tensor that helper expects rather than a pre-split per-level tuple to reuse existing `deformable_attention_core_func_v2` in `utils.py` without modification.
- making `bbox_head` and `pre_bbox_head` FP32. See 
  - https://github.com/Peterande/D-FINE/issues/3, 
  - https://github.com/Peterande/D-FINE/issues/43, and 
  - https://github.com/Peterande/D-FINE/issues/199 in D-FINE repo and 
  - #707 in our repo. 
  I also created an issue for potential ONNX export failure.

Note that currently:
- this does not include the necessary changes for the D-FINE losses. Will add in a follow-up PR.
- the default decoder is the original RTDETRv2 decoder. Will switch once we conclude the loss implementation and get proper results from benchmarks.

## How has it been tested?

- N/A (not in any testable state)

## Did you update [CHANGELOG.md](../CHANGELOG.md)?

- [ ] Yes
- [x] Not needed (to be added after finishing D-FINE loss)

## Did you update the documentation?

- [ ] Yes
- [x] Not needed (internal change without effects for user)

### CODE REVIEW

File: src/lightly_train/_task_models/dinov2_ltdetr_object_detection/train_model.py [hunk 1]
```diff
@@ -75,6 +75,7 @@ class DINOv2LTDETRObjectDetectionTrainArgs(TrainModelArgs):
     backbone_url: str = ""
     backbone_args: dict[str, Any] = {}
     backbone_freeze: bool = False
+    decoder: Literal["rtdetrv2", "dfine"] = "rtdetrv2"
```
Comment (liopeer) [inline src/lightly_train/_task_models/dinov2_ltdetr_object_detection/train_model.py]: Suggestion: Name it `decoder_name` (in accordance with `model_name`).

File: src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py [hunk 2]
```diff
@@ -41,6 +44,38 @@
 
 logger = logging.getLogger(__name__)
 
+_LTDETRDecoder = Literal["rtdetrv2", "dfine"]
+
+
+def _build_decoder(
```
Comment (liopeer) [inline src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py]: Nit: I prefer to have private functions at the bottom of a module.

File: src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py [hunk 3]
```diff
@@ -212,6 +247,50 @@ class _RTDETRTransformerv2ViTGConfig(_RTDETRTransformerv2Config):
     dim_feedforward: int = 12288
 
 
+class _DFINETransformerConfig(PydanticConfig):
```
Comment (liopeer) [inline src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py:220]: Did not check all of these values. Assuming you copied them from the configs of the RTDETR decoder that are just above.

Comment (liopeer) [reply src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py:220]: Actually, is it deliberate that we have `num_layers=6`?

File: src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py [hunk 4]
```diff
@@ -41,6 +44,38 @@
...
+    *,
+    config: _DINOv2LTDETRObjectDetectionConfig,
+    decoder: _LTDETRDecoder,
+    num_classes: int,
+    image_size: tuple[int, int],
+    cross_attn_method: str | None = None,
+) -> RTDETRTransformerv2 | DFINETransformer:
+    if decoder == "rtdetrv2":
+        decoder_config = config.rtdetr_transformer.model_dump()
+        if cross_attn_method is not None:
```
Comment (liopeer) [inline src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py]: Is this actually even possible? Should it not throw an error?

File: src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py [hunk 5]
```diff
@@ -212,6 +247,50 @@ class _RTDETRTransformerv2ViTGConfig(_RTDETRTransformerv2Config):
...
+    num_layers: int = 6
+    num_queries: int = 300
+    num_denoising: int = 100
+    label_noise_ratio: float = 0.5
+    box_noise_scale: float = 1.0
+    eval_idx: int = -1
+    num_points: list[int] = [3, 6, 3]
+    query_select_method: str = "default"
+    cross_attn_method: str = "default"
+    dim_feedforward: int = 1024
```
Comment (liopeer) [inline src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py]: This is very small for the big models, no?

File: src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py [hunk 6]
```diff
@@ -373,6 +467,7 @@ def __init__(
         config_name = config_name.replace("-noreg", "")
         config_cls = config_mapping[config_name]
         config = config_cls()
+        config.decoder = decoder
```
Comment (liopeer) [inline src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py]: Another reason to use `decoder_name` and not `decoder`. I am surprised that mypy does not afterwards complain, because here it's typed as `config.decoder: str` and in line 486 it will be typed as `config.decoder: RTDETRTransformerv2 | DFINETransformer`.

File: src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py [hunk 7]
```diff
@@ -782,6 +878,7 @@ def __init__(
         config_name = parsed_name["backbone_name"]
         config_cls = config_mapping[config_name]
         config = config_cls()
+        config.decoder = decoder
```
Comment (liopeer) [inline src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py]: Same here ofc.

File: src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py [hunk 8]
```diff
@@ -36,6 +36,9 @@
     DINOv3STAs,
 )
 from lightly_train._task_models.object_detection_components import tiling_utils
+from lightly_train._task_models.object_detection_components.dfine_decoder import (
```
Comment (liopeer) [inline src/lightly_train/_task_models/dinov3_ltdetr_object_detection/task_model.py:39]: Same points as in the DINOv2 variant.

File: src/lightly_train/_task_models/object_detection_components/hooks.py [hunk 9]
```diff
@@ -0,0 +1,228 @@
+#
```
Comment (liopeer) [inline src/lightly_train/_task_models/object_detection_components/hooks.py:1]: AFAIK this module is now entirely from us, right? If that's correct we should not have it under Apache 2.0, i.e. we should adjust the the Makefile. I would actually also include another license file for D-FINE (and another header), even though RT-DETRv2 and D-FINE are both Apache 2.0. It just then makes it clear from the Makefile which module was vendored from which other project.

File: src/lightly_train/_task_models/object_detection_components/hooks.py [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+import logging
+from typing import Any, Literal
+
+import torch
+from torch.nn import Module, ModuleList
+
+logger = logging.getLogger(__name__)
+
+
+def denoising_class_embed_reuse_or_reinit_hook(
```
Comment (liopeer) [inline src/lightly_train/_task_models/object_detection_components/hooks.py:31]: Assuming all these functions are directly copied.

File: src/lightly_train/_task_models/object_detection_components/rtdetrv2_decoder.py [hunk 11]
```diff
@@ -12,39 +12,40 @@
...
 
-# Modifications Copyright 2025 Lightly AG:
-# - added load state dict pre hooks to reinitialize the classification score heads
-#   and denoising class embedding if the number of classes has changed
-# - implemented `score_head_reuse_or_reinit_hook`, `_score_head_reuse_or_reinit_hook`,
-#   `_reuse_or_reinit`, and `denoising_class_embed_reuse_or_reinit_hook` functions
+# Modifications Copyright 2026 Lightly AG:
+#  - Registered ``score_head_reuse_or_reinit_hook`` and (when denoising is
+#     enabled) ``denoising_class_embed_reuse_or_reinit_hook`` as
+#     load-state-dict pre-hooks, so pretrained D-FINE checkpoints can be
```
Comment (liopeer) [inline src/lightly_train/_task_models/object_detection_components/rtdetrv2_decoder.py]:
```suggestion
#     load-state-dict pre-hooks, so pretrained RT-DETR checkpoints can be
```

File: src/lightly_train/_task_models/object_detection_components/hooks.py (hunk 9 above)
Comment (liopeer) [reply src/lightly_train/_task_models/object_detection_components/hooks.py:1]: We also have a separate license file for e.g. TIMM, which is also Apache 2.0

File: src/lightly_train/_task_models/object_detection_components/dfine_utils.py [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+
+import torch
+from torch import Tensor
+
+from lightly_train._task_models.object_detection_components.box_ops import (
+    box_xyxy_to_cxcywh,
+)
+
+
+def weighting_function(
```
Comment (liopeer) [inline src/lightly_train/_task_models/object_detection_components/dfine_utils.py:28]: Not checking, since I assume it's directly from D-FINE.

**Review by liopeer — COMMENTED** [review]
Overall looking solid. Left a few comments for smaller improvements!

File: src/lightly_train/_task_models/object_detection_components/hooks.py (hunk 9 above)
Comment (yutong-xiang-97) [reply src/lightly_train/_task_models/object_detection_components/hooks.py:1]: These hooks are by us entirely, but yes we need to include D-FINE in NOTICE and Makefile.

File: src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py (hunk 5 above)
Comment (yutong-xiang-97) [reply src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py]: For decoder's `dim_feedforward` D-FINE uses 1024 across size S to X. For hybrid encoder, it does use 2048 for X.

That being said I will use DEIMv2 configs anyway and that DEIMv2 decoder's `dim_feedforward` does scale with the model size.

File: src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py (hunk 3 above)
Comment (yutong-xiang-97) [reply src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py:220]: Ah good point! This is different across the model size. With small models this is 3. Will fix.

Comment (yutong-xiang-97) [reply src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py:220]: Actually we may need to use DEIMv2's default here since D-FINE only uses HGNetv2 as backbone. DEIMv2's decoder also uses D-FINE's decoder and losses.

Comment (liopeer) [reply src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py:220]: Now we have `num_layers=4` everywhere except for the super class. I am a bit confused.

File: src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py [hunk 13]
```diff
@@ -212,6 +217,58 @@ class _RTDETRTransformerv2ViTGConfig(_RTDETRTransformerv2Config):
...
+    num_layers: int = 6
+    num_queries: int = 300
+    num_denoising: int = 100
+    label_noise_ratio: float = 0.5
+    box_noise_scale: float = 1.0
+    eval_idx: int = -1
+    num_points: list[int] = [3, 6, 3]
+    query_select_method: str = "default"
+    cross_attn_method: str = "default"
+    dim_feedforward: int = 2048
```
Comment (liopeer) [inline src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py]: Would remove the default if it's something else in all the models. Easy source of errors.

Comment (liopeer) [reply src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py]: Same applies to `num_layers` (if different across model sizes).

File: src/lightly_train/_task_models/object_detection_components/hooks.py (hunk 9 above)
Comment (liopeer) [reply src/lightly_train/_task_models/object_detection_components/hooks.py:1]: Seems to still be Apache 2.0 licensed.

**Review by liopeer — COMMENTED** [review]
Overall LGTM, but I am not sure if every comment was addressed that I previously raised. Would be great to quickly get answers to this before full approval.

File: src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py (hunk 4 above)
Comment (yutong-xiang-97) [reply src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py]: Could you point me to where the error could occur? The training runs without errors and `_RTDETRTransformerv2Config` is fully serializable.

File: src/lightly_train/_task_models/object_detection_components/dfine_decoder.py [hunk 14]
```diff
@@ -0,0 +1,1063 @@
...
+    ):
+        """Multi-Scale Deformable Attention"""
+        super().__init__()
+        self.embed_dim = embed_dim
+        self.num_heads = num_heads
+        self.num_levels = num_levels
+        self.offset_scale = offset_scale
+
+        if isinstance(num_points, list):
+            assert len(num_points) == num_levels, ""
```
Comment (simonschoelly) [inline src/lightly_train/_task_models/object_detection_components/dfine_decoder.py:101]: What is the purpose of emitting "" here if the assert fails?

Comment (yutong-xiang-97) [reply src/lightly_train/_task_models/object_detection_components/dfine_decoder.py:101]: This is from the original d-fine code, I would just leave it as-is.

**Review by liopeer — APPROVED** [review]
LGTM! But leave @simonschoelly to resolve his comment on his side.
